### PR TITLE
add box clip transforms

### LINKED ISSUE #3572: Box transforms
**Is your feature request related to a problem? Please describe.**
Box pre- and post-processing transform functions are not in MONAI

**Describe the solution you'd like**
Add box transform functions

**Describe alternatives you've considered**
N/A

**Additional context**
1. convert input box with original mode to the box with standard mode
2. convert output box with standard mode to the box with original mode
3. Flip boxes with image as a transform

### PR DESCRIPTION
**Pull request by Can-Zhao**
Signed-off-by: Can Zhao <canz@nvidia.com>

Fixes #3572 .

### Description
add box clip transforms

### Status
**Ready/Work in progress/Hold**

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [x] New tests added to cover the changes.
- [ ] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [ ] Quick tests passed locally by running `./runtests.sh --quick --unittests  --disttests`.
- [x] In-line docstrings updated.
- [ ] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

File: monai/apps/detection/transforms/array.py [hunk 1]
```diff
@@ -296,3 +305,61 @@ def __call__(  # type: ignore
         """
 
         return flip_boxes(boxes, spatial_size=spatial_size, flip_axes=self.spatial_axis)
+
+
+class ClipBoxToImage(Transform):
```
Comment (Nic-Ma) [inline monai/apps/detection/transforms/array.py:310]: Please add the new transforms into `__all__` list:
https://github.com/Project-MONAI/MONAI/blob/dev/monai/apps/detection/transforms/array.py#L28

Thanks.

File: monai/apps/detection/transforms/array.py [hunk 2]
```diff
@@ -296,3 +305,61 @@ def __call__(  # type: ignore
...
+
+        labels_tuple = ensure_tuple(labels)
+        labels_clip_list = []
+
+        keep_t: torch.Tensor = convert_data_type(keep, torch.Tensor)[0]
+        for i in range(len(labels_tuple)):
+            labels_t: torch.Tensor = convert_data_type(labels_tuple[i], torch.Tensor)[0]
+            if boxes.shape[0] != labels_t.shape[0]:
+                raise ValueError("boxes.shape[0] should be equal to the number of labels or scores.")
+            labels_t = deepcopy(labels_t)[keep_t, ...]
```
Comment (Nic-Ma) [inline monai/apps/detection/transforms/array.py:363]: @wyli I am not very sure whether `labels_t = deepcopy(labels_t[keep_t, ...])` would be better?

Thanks.

File: monai/apps/detection/transforms/dictionary.py [hunk 3]
```diff
@@ -687,10 +697,81 @@ def inverse(self, data: Mapping[Hashable, NdarrayOrTensor]) -> Dict[Hashable, Nd
...
+        d[self.box_keys], clipped_labels = self.clipper(d[self.box_keys], labels, spatial_size)
+
+        for label_key, clipped_labels_i in zip(self.label_keys, clipped_labels):
+            self.push_transform(d, label_key)
+            d[label_key] = clipped_labels_i
+        return d
+
+    def inverse(self, data: Mapping[Hashable, NdarrayOrTensor]) -> Dict[Hashable, NdarrayOrTensor]:
+        d = deepcopy(dict(data))
+        # clipped boxes cannot be restored
```
Comment (Nic-Ma) [inline monai/apps/detection/transforms/dictionary.py:764]: If it doesn't support inverse, please don't inherit from `InvertibleTransform`.

Thanks.

File: monai/apps/detection/transforms/dictionary.py [hunk 4]
```diff
@@ -687,10 +697,81 @@ def inverse(self, data: Mapping[Hashable, NdarrayOrTensor]) -> Dict[Hashable, Nd
...
+class ClipBoxToImaged(MapTransform, InvertibleTransform):
+    """
+    Dictionary-based wrapper of :py:class:`monai.apps.detection.transforms.array.ClipBoxToImage`.
+
+    Clip the bounding boxes and the associated labels/scores to makes sure they are within the image.
+    There might be multiple keys of labels/scores associated with one key of boxes.
+
+    Args:
+        box_keys: The single key to pick box data for transformation. The box mode is assumed to be ``StandardMode``.
+        label_keys: Keys that represents the lables corresponding to the ``box_keys``.
```
Comment (Nic-Ma) [inline monai/apps/detection/transforms/dictionary.py:709]: Do you support multiple `label_keys`?
I feel it only supports single key as `box_keys`.

Thanks.

File: monai/apps/detection/transforms/dictionary.py [hunk 5]
```diff
@@ -687,10 +697,81 @@ def inverse(self, data: Mapping[Hashable, NdarrayOrTensor]) -> Dict[Hashable, Nd
...
+        self.clipper = ClipBoxToImage(remove_empty=remove_empty)
+
+    def __call__(self, data: Mapping[Hashable, NdarrayOrTensor]) -> Dict[Hashable, NdarrayOrTensor]:
+        d = dict(data)
+        spatial_size = d[self.box_ref_image_keys].shape[1:]
+        self.push_transform(d, self.box_keys)
+        labels = [d[label_key] for label_key in self.label_keys]
+        d[self.box_keys], clipped_labels = self.clipper(d[self.box_keys], labels, spatial_size)
+
+        for label_key, clipped_labels_i in zip(self.label_keys, clipped_labels):
```
Comment (Nic-Ma) [inline monai/apps/detection/transforms/dictionary.py:757]: So here it should only have 1 item, let's remove the for-loop?

Thanks.

File: monai/apps/detection/transforms/dictionary.py (hunk 4 above)
Comment (Can-Zhao) [reply monai/apps/detection/transforms/dictionary.py:709]: Yes, it supports multiple label_keys. I put an example in docstring.

File: monai/apps/detection/transforms/dictionary.py (hunk 5 above)
Comment (Can-Zhao) [reply monai/apps/detection/transforms/dictionary.py:757]: There might be multiple items

**Review by Nic-Ma — APPROVED** [review]
Thanks for the quick update.
Looks good to me.

Thanks.

**Pull request merged**